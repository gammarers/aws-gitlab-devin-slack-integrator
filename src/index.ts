import { Duration, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { HttpApi, HttpMethod } from 'aws-cdk-lib/aws-apigatewayv2';
import * as integrations from 'aws-cdk-lib/aws-apigatewayv2-integrations';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';
import { IntegratorFunction } from './funcs/integrator-function';

export interface GitLabDevinSlackIntegratorProps {
  readonly secretName: string;
}

export class GitLabDevinSlackIntegrator extends Construct {
  constructor(scope: Construct, id: string, props: GitLabDevinSlackIntegratorProps) {
    super(scope, id);

    // üëá Get account & region
    const account = Stack.of(scope).account;
    const region = Stack.of(scope).region;
    const partition = Stack.of(scope).partition;

    // Create DynamoDB(GitLab MR ID & Slack message timestamp)
    const table = new dynamodb.Table(this, 'GitLabMargeRquestTimestampTable', {
      partitionKey: { name: 'mrid', type: dynamodb.AttributeType.NUMBER },
      removalPolicy: RemovalPolicy.DESTROY,
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
    });

    // GitLab Webhook Lambda
    const messangerFunction = new IntegratorFunction(this, 'IntegratorFunction', {
      description: 'Send Slack message function',
      architecture: lambda.Architecture.ARM_64,
      role: new iam.Role(this, 'MessangerFunctionExecutionRole', {
        // roleName: names.functionRoleName,
        description: '',
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
        ],
        inlinePolicies: {
          'get-secret-policy': new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  'secretsmanager:GetSecretValue',
                ],
                resources: [
                  `arn:${partition}:secretsmanager:${region}:${account}:secret:${props.secretName}*`,
                ],
              }),
            ],
          }),
        },
      }),
      timeout: Duration.seconds(10),
      paramsAndSecrets: lambda.ParamsAndSecretsLayerVersion.fromVersion(lambda.ParamsAndSecretsVersions.V1_0_103, {
        cacheSize: 500,
        logLevel: lambda.ParamsAndSecretsLogLevel.INFO,
      }),
      environment: {
        TABLE_NAME: table.tableName,
        SECRET_NAME: props.secretName,
      },
      loggingFormat: lambda.LoggingFormat.JSON,
      systemLogLevelV2: lambda.SystemLogLevel.INFO,
      applicationLogLevelV2: lambda.ApplicationLogLevel.INFO,
      logGroup: new logs.LogGroup(this, 'MessangerFunctionLogGroup', {
        //logGroupName: names.functionName ? `/aws/lambda/${names.functionName}` : undefined,
        retention: logs.RetentionDays.THREE_MONTHS,
        removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      }),
    });
    // todo: get/put/delete„ÅÆÂé≥ÂØÜ„Å™ÂÆöÁæ©„ÇíË°å„ÅÜ
    table.grantReadWriteData(messangerFunction);

    // HTTP API
    const api = new HttpApi(this, 'SlackGitlabIntegrationAPI', {
      apiName: 'Slack GitLab Integration API',
    });

    // GitLab Webhook Áî®„É´„Éº„ÉàÔºàpath-based channelÔºâ
    api.addRoutes({
      path: '/gitlab/merge/request/{channel}',
      methods: [HttpMethod.POST],
      integration: new integrations.HttpLambdaIntegration('GitlabWebhookIntegration', messangerFunction),
    });
  }
}