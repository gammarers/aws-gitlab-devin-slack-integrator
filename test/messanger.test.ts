import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GitLabMergeRequestDevinReviewSlackMessanger } from '../src';

describe('Subscription Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  new GitLabMergeRequestDevinReviewSlackMessanger(stack, 'GitLabMergeRequestDevinReviewSlackMessanger', {
    secretName: 'foobar',
  });

  const template = Template.fromStack(stack);

  it('Is Lambda Function Resource Count', () => {
    template.resourceCountIs('AWS::Lambda::Function', 1);
  });

  it('Is IAM Role Resource Count', () => {
    template.resourceCountIs('AWS::IAM::Role', 1);
  });

  it('Is Logs LogGroup Resource Count', () => {
    template.resourceCountIs('AWS::Logs::LogGroup', 1);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});