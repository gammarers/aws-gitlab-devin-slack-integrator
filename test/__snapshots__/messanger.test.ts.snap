// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Testing Should match snapshot 1`] = `
{
  "Mappings": {
    "ParamsandsecretslayerMap": {
      "af-south-1": {
        "1x0x103xarm64": "arn:aws:lambda:af-south-1:317013901791:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "ap-east-1": {
        "1x0x103xarm64": "arn:aws:lambda:ap-east-1:768336418462:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "ap-northeast-1": {
        "1x0x103xarm64": "arn:aws:lambda:ap-northeast-1:133490724326:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "ap-northeast-2": {
        "1x0x103xarm64": "arn:aws:lambda:ap-northeast-2:738900069198:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "ap-south-1": {
        "1x0x103xarm64": "arn:aws:lambda:ap-south-1:176022468876:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "ap-southeast-1": {
        "1x0x103xarm64": "arn:aws:lambda:ap-southeast-1:044395824272:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "ap-southeast-2": {
        "1x0x103xarm64": "arn:aws:lambda:ap-southeast-2:665172237481:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "ap-southeast-3": {
        "1x0x103xarm64": "arn:aws:lambda:ap-southeast-3:490737872127:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "ca-central-1": {
        "1x0x103xarm64": "arn:aws:lambda:ca-central-1:200266452380:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "eu-central-1": {
        "1x0x103xarm64": "arn:aws:lambda:eu-central-1:187925254637:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "eu-north-1": {
        "1x0x103xarm64": "arn:aws:lambda:eu-north-1:427196147048:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "eu-south-1": {
        "1x0x103xarm64": "arn:aws:lambda:eu-south-1:325218067255:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "eu-west-1": {
        "1x0x103xarm64": "arn:aws:lambda:eu-west-1:015030872274:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "eu-west-2": {
        "1x0x103xarm64": "arn:aws:lambda:eu-west-2:133256977650:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "eu-west-3": {
        "1x0x103xarm64": "arn:aws:lambda:eu-west-3:780235371811:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "me-south-1": {
        "1x0x103xarm64": "arn:aws:lambda:me-south-1:832021897121:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "sa-east-1": {
        "1x0x103xarm64": "arn:aws:lambda:sa-east-1:933737806257:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "us-east-1": {
        "1x0x103xarm64": "arn:aws:lambda:us-east-1:177933569100:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "us-east-2": {
        "1x0x103xarm64": "arn:aws:lambda:us-east-2:590474943231:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
      "us-west-1": {
        "1x0x103xarm64": "arn:aws:lambda:us-west-1:997803712105:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:1",
      },
      "us-west-2": {
        "1x0x103xarm64": "arn:aws:lambda:us-west-2:345057560386:layer:AWS-Parameters-and-Secrets-Lambda-Extension-Arm64:4",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GitLabMergeRequestDevinReviewSlackMessangerGitLabMargeRquestTimestampTable52201E76": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "mrid",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "mrid",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunction0486EA84": {
      "DependsOn": [
        "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleDefaultPolicy7906E3A5",
        "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleF1483720",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "558e55ee38395b00763d8be2467dd1d59e61218a7728dd78df0db5595de34041.zip",
        },
        "Description": "Send Slack message function",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "PARAMETERS_SECRETS_EXTENSION_CACHE_ENABLED": "true",
            "PARAMETERS_SECRETS_EXTENSION_CACHE_SIZE": "500",
            "PARAMETERS_SECRETS_EXTENSION_HTTP_PORT": "2773",
            "PARAMETERS_SECRETS_EXTENSION_LOG_LEVEL": "info",
            "PARAMETERS_SECRETS_EXTENSION_MAX_CONNECTIONS": "3",
            "SECRETS_MANAGER_TIMEOUT_MILLIS": "0",
            "SECRETS_MANAGER_TTL": "300",
            "SECRET_NAME": "foobar",
            "SSM_PARAMETER_STORE_TIMEOUT_MILLIS": "0",
            "SSM_PARAMETER_STORE_TTL": "300",
            "TABLE_NAME": {
              "Ref": "GitLabMergeRequestDevinReviewSlackMessangerGitLabMargeRquestTimestampTable52201E76",
            },
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Fn::FindInMap": [
              "ParamsandsecretslayerMap",
              {
                "Ref": "AWS::Region",
              },
              "1x0x103xarm64",
            ],
          },
        ],
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionLogGroup9C69643E",
          },
          "SystemLogLevel": "INFO",
        },
        "Role": {
          "Fn::GetAtt": [
            "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleF1483720",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleDefaultPolicy7906E3A5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "GitLabMergeRequestDevinReviewSlackMessangerGitLabMargeRquestTimestampTable52201E76",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleDefaultPolicy7906E3A5",
        "Roles": [
          {
            "Ref": "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleF1483720",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionExecutionRoleF1483720": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "secretsmanager:GetSecretValue",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":secretsmanager:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":secret:foobar*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "get-secret-policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunctionLogGroup9C69643E": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "RetentionInDays": 90,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPI2BAAC04C": {
      "Properties": {
        "Name": "Slack GitLab Integration API",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPIDefaultStage09489DC7": {
      "Properties": {
        "ApiId": {
          "Ref": "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPI2BAAC04C",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPIPOSTgitlabmergerequestchannelA0220D6B": {
      "Properties": {
        "ApiId": {
          "Ref": "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPI2BAAC04C",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /gitlab/merge/request/{channel}",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPIPOSTgitlabmergerequestchannelGitlabWebhookIntegrationA1B552DC",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPIPOSTgitlabmergerequestchannelGitlabWebhookIntegrationA1B552DC": {
      "Properties": {
        "ApiId": {
          "Ref": "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPI2BAAC04C",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunction0486EA84",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPIPOSTgitlabmergerequestchannelGitlabWebhookIntegrationPermission7EBDC253": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GitLabMergeRequestDevinReviewSlackMessangerMessangerFunction0486EA84",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "GitLabMergeRequestDevinReviewSlackMessangerSlackGitlabIntegrationAPI2BAAC04C",
              },
              "/*/*/gitlab/merge/request/{channel}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
